MCH: refactoring of MCH code

aferrero2707 opened the pull request.

Barthelemy commented:
Hi, 
Is it something you still want to do or should I close this PR ? 
Cheers
Barth

aferrero2707 commented:
@aphecetche @Barthelemy @chiarazampolli this large PR contains lots of changes to the MCH QC code, needed to properly introduce post-processing and trends as well as data quality processing.

The code is in principle ready to be merged, and all checks are green. Ideally I would like to merge it as soon as possible so that we can test it next week, but we should absolutely avoid that the code is included in the O2 version for apass3.

How do you suggest to proceed? Thanks!

Barthelemy commented:
I will review today but I see that there are conflicts. Could you take care of them ? 
I don't know how to proceed to avoid having it for apass3.

aphecetche commented:
The conflicts are due to the change of my PR #1641. @aferrero2707 I can take of the conflicts and push an updated version

aferrero2707 commented:
@aphecetche the three conflicting header files have disappeared in my PR. If it is easier I can rebase my code and fix the conflicts later in the afternoon, it should still allow to have the CI results before tomorrow.

aphecetche commented:
@aferrero2707 ok, but you'll see there's also some tweaking to be done in the TracksTask to get the geometry transformation correctly... 
I guess at some point we should make a ClusterPlotter that would take responsibility for all cluster related things (all but the number of cluster per track) and remove those parts from TrackPlotter.

aferrero2707 commented:
@aphecetche done. The MCH TrackTask code is now exactly the one of your last PR. I had introduced a plotter class in my branch, but this part needs to be reviewed anyhow so I reverted all changes. As we already discussed, we need to prepare a dedicated PR to harmonize the MCH and MUON tracks QC and avoid code duplication.

aphecetche commented:
@aferrero2707 ok, thanks. Are all the configuration files under `MCH/etc/` supposed to be usable as-is ? I've briefly tried `qc-mch-digits.json` and got :
```
[28742:PP-TASK-RUNNER-Digits]: std::exception::what: No such node (qc.postprocessing.Digits.plots)
```

aferrero2707 commented:
@aphecetche yes, they are, but they clearly contained some mistakes... they are fixed now. I tried both the digits and preclusters ones locally and they worked after fixing the mistakes.

aphecetche commented:
@aferrero2707 could you share the command(s) you use to run the digits QC for instance ?

Trying something like : 

```
o2-mch-sim-digits-reader-workflow --disable-mc --hbfutils-config o2_tfidinfo.root | o2-qc --config json://../QualityControl/Modules/MUON/MCH/etc/qc-mch-digits.json --override-values "qc.config.database.host=localhost:6464;qc.config.database.maxObjectSize=2342912;qc.config.conditionDB=localhost:6464"
```

(where localhost:6464 is my local CCDB), I get : 

```
[54878:PP-TASK-RUNNER-Digits]: [14:26:02][ERROR] Requested resource does not exist: localhost:6464/qc/MCH/MO/Digits/Occupancy_Elec/1676467562/
[54878:PP-TASK-RUNNER-Digits]: 2023-02-15 14:26:02.696498 !!! Error - We could NOT retrieve the object qc/MCH/MO/Digits/Occupancy_Elec with timestamp 1676467562.
```

and the process never ends. The `Occupancy_Elec` monitoring object does exist in the CCDB, but with a validity end strictly equals to validity start, and so the requested time is indeed not in the validity range : is that expected ?

Note also I had to increase the maxObjectSize, as the default of 2MB is not enough. On a next PR we should really use a contiguous DSID and gain about a factor 2 on those (whatever vs DSID) 2D histos...

aferrero2707 commented:
@aphecetche the error you see at task initialization corresponds to the attempted retrieval of the reference plot, for which I have set a dummy date in the configuration. Afterwards you should see that the objects are retrieved successfully, something like this:
```
[19463:PP-TASK-RUNNER-Digits]: [16:56:25][ERROR] Requested resource does not exist: http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/Occupancy_Elec/1676476585/
[19463:PP-TASK-RUNNER-Digits]: 2023-02-15 16:56:25.484699 !!! Error - We could NOT retrieve the object qc/MCH/MO/Digits/Occupancy_Elec with timestamp 1676476585.
[19463:PP-TASK-RUNNER-Digits]: 2023-02-15 16:56:25.484716 !!! Error - An error occurred during retrieval
[19463:PP-TASK-RUNNER-Digits]: [16:56:25][ERROR] Requested resource does not exist: http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/OccupancySignal_Elec/1676476585/
[19463:PP-TASK-RUNNER-Digits]: 2023-02-15 16:56:25.491254 !!! Error - We could NOT retrieve the object qc/MCH/MO/Digits/OccupancySignal_Elec with timestamp 1676476585.
[19463:PP-TASK-RUNNER-Digits]: 2023-02-15 16:56:25.491267 !!! Error - An error occurred during retrieval
[19463:PP-TASK-RUNNER-Digits]: [16:56:35][INFO] Correctly handshaken websocket connection.
[19463:PP-TASK-RUNNER-Digits]: 2023-02-15 16:57:10.522098     Updating the user task due to trigger 'triggerType: 7, timestamp: 1676476625298'
[19463:PP-TASK-RUNNER-Digits]: [16:57:10][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/Occupancy_Elec/1676476605856/59a6dda7-ad49-11ed-8dee-c0a80209250c for 1676476625298 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
[19463:PP-TASK-RUNNER-Digits]: [16:57:18][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/OccupancySignal_Elec/1676476605856/5a204c89-ad49-11ed-8dee-c0a80209250c for 1676476625298 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
[19463:PP-TASK-RUNNER-Digits]: [16:57:26][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/DigitOrbitInTF/1676476605856/5a9d809d-ad49-11ed-8dee-c0a80209250c for 1676476625298 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
[19463:PP-TASK-RUNNER-Digits]: [16:57:29][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/DigitSignalOrbitInTF/1676476605856/5b075b74-ad49-11ed-8dee-c0a80209250c for 1676476625298 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
[19463:PP-TASK-RUNNER-Digits]: 2023-02-15 16:57:41.011855     Updating the user task due to trigger 'triggerType: 7, timestamp: 1676476655297'
[19463:PP-TASK-RUNNER-Digits]: [16:57:41][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/Occupancy_Elec/1676476605856/59a6dda7-ad49-11ed-8dee-c0a80209250c for 1676476655297 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
[19463:PP-TASK-RUNNER-Digits]: [16:57:41][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/OccupancySignal_Elec/1676476605856/5a204c89-ad49-11ed-8dee-c0a80209250c for 1676476655297 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
[19463:PP-TASK-RUNNER-Digits]: [16:57:42][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/DigitOrbitInTF/1676476605856/5a9d809d-ad49-11ed-8dee-c0a80209250c for 1676476655297 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
[19463:PP-TASK-RUNNER-Digits]: [16:57:42][INFO] ccdb reads http://ccdb-test.cern.ch:8080/qc/MCH/MO/Digits/DigitSignalOrbitInTF/1676476636334/6cf71515-ad49-11ed-8dee-c0a80209250c for 1676476655297 (retrieve, agent_id: mchflp1-1676476585-roIkGQ), 
```

Your command looks ok. This is what I am using to test the code with CTF files:
```
o2-ctf-reader-workflow --session default --shm-segment-size 16000000000 --ctf-input "ctflist-529663.txt" --remote-regex "^alien://.+" --copy-cmd no-copy --onlyDet MCH --max-tf -1 --delay 0 | o2-tfidinfo-writer-workflow --session default --shm-segment-size 16000000000 | o2-mch-reco-workflow --session default --shm-segment-size 16000000000 --disable-mc --disable-root-input --disable-root-output --digits --configKeyValues "MCHDigitFilter.rejectBackground=true;MCHDigitFilter.timeOffset=120;MCHDigitFilter.minADC=1;MCHTimeClusterizer.onlyTrackable=true;MCHTimeClusterizer.peakSearchSignalOnly=true;MCHClustering.lowestPadCharge=10;MCHClustering.defaultClusterResolution=0.4;MCHTracking.chamberResolutionX=0.4;MCHTracking.chamberResolutionY=0.4;MCHTracking.sigmaCutForTracking=7.;MCHTracking.sigmaCutForImprovement=6." | o2-qc --session default --shm-segment-size 16000000000 --config json://home/flp/Workflows/qc.json | o2-dpl-run --session default --shm-segment-size 16000000000 --batch --run
```

aphecetche commented:
@aferrero2707 I'm afraid you have a couple of conflicts to resolve, following the merging of #1656. Sorry about that.

aphecetche commented:
@aferrero2707 two other points to consider, in the `MUON/Common/src/TrackPlotter.cxx` : 

- the `mTrackChi2OverNDF` histograms are created but not filled as far as I can tell
- the `fillTrackPairHistos` should be called with the list of tracks per ROF, not per TF (see e.g. in the "old" task  https://github.com/AliceO2Group/QualityControl/blob/496c9208bf52dd8a5ee81f6ddf85a2cb77c3c8b4/Modules/MUON/MCH/src/TracksTask.cxx#L259-L264)

aferrero2707 commented:
> @aferrero2707 two other points to consider, in the `MUON/Common/src/TrackPlotter.cxx` :
> 
>     * the `mTrackChi2OverNDF` histograms are created but not filled as far as I can tell

I propose to remove them from the global tracks, because I would not know exactly which Chi2 to use to fill the histogram in this case

> 
>     * the `fillTrackPairHistos` should be called with the list of tracks per ROF, not per TF (see e.g. in the "old" task  https://github.com/AliceO2Group/QualityControl/blob/496c9208bf52dd8a5ee81f6ddf85a2cb77c3c8b4/Modules/MUON/MCH/src/TracksTask.cxx#L259-L264
>       )

The code currently loops over all the track combinations and only combines tracks that are less than 1us apart in time (see [here](https://github.com/AliceO2Group/QualityControl/blob/8dc17a43438bf47b2a3ad7d499c6c0dd6f8036d9/Modules/MUON/Common/src/TrackPlotter.cxx#L118)). This should be equivalent to requiring that they belong to the same ROF, but more flexible. However there might be something I have overlooked... The 1us cut is probably too conservative, and could be reduced to ~500ns, but seemed to be sufficiently selective in my local tests.

However we can revert back to the ROF-based selection if we have doubts about the time cut.

aferrero2707 commented:
@aphecetche @Barthelemy @knopers8 I have re-organized the commits in a more coherent way, hopefully this will make the review of this very large PR a bit more comfortable and easy.

aferrero2707 commented:
@Barthelemy @knopers8 apart from the various improvements, with this new version of the code the distinct number of objects per run is reduced to ~150, instead of the >600 objects of the current MCH QC.

Barthelemy commented:
@TimoWilken @ktf there is no space left on the macos build machine: `fatal error: error in backend: IO failure on output stream: No space left on device`

Barthelemy commented:
I went through the changes and did not spot any particular problem. Given the size of the PR I could not make an in-depth review. 

Do you want to merge it now ? I presume that it will entail quite many changes in the config files at P2 and perhaps in the JIT workflows ?

aferrero2707 commented:
@Barthelemy thanks! I understand that the amount of changes are such that an in-depth review would not be possible.
I would be in favor of merging this week, such that we have enough buffer to test the new code at P2 and eventually implement bugfixes before the stable beams come back.

regarding the update of the QC configurations at P2, do we still have to modify the ControlWorkflows, or the Consul files are enough since we use JIT? In any case I will take care of the ControlWorkflows once the PR is merged.

@aphecetche do you have any pending checks before we merge?

aferrero2707 commented on Modules/MUON/Common/src/MergeableTH2Ratio.cxx:
```diff
@@ -147,4 +161,80 @@ void MergeableTH2Ratio::Reset(Option_t* option)
+                                Int_t ny, Double_t ymin, Double_t ymax,
+                                Int_t nz, Double_t zmin, Double_t zmax)
+{
+  std::cout << "SetBins(Int_t nx, Double_t xmin, Double_t xmax, Int_t ny, Double_t ymin, Double_t ymax, Int_t nz, Double_t zmin, Double_t zmax) not valid for MergeableTH2Ratio" << std::endl;
```
Thanks for spotting this, I replaced std::cout by ILOG(Debug, Devel).
